## Morgan Southgate    
## Ecological Genomics:   

### Overall Description of notebook      

This notebook will catalog my entire work throughout the semester, including all my relevant code, results, and interpretations


### Date started: January 24, 2018   
### Date end:   (year-month-day)    

### Philosophy   
Science should be reproducible and one of the best ways to achieve this is by logging research activities in a notebook. Because science/biology has increasingly become computational, it is easier to document computational projects in an electronic form, which can be shared online through Github.    

### Helpful features of the notebook     

**It is absolutely critical for your future self and others to follow your work.**     

* The notebook is set up with a series of internal links from the table of contents.    
* All notebooks should have a table of contents which has the "Page", date, and title (information that allows the reader to understand your work).     
* Also, one of the perks of keeping all activities in a single document is that you can **search and find elements quickly**.     
* Lastly, you can share specific entries because of the three "#" automatically creates a link when the notebook renders on github.      


<a rel="license" href="http://creativecommons.org/licenses/by/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by/4.0/88x31.png" /></a><br />This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution 4.0 International License</a>.  


### Table of contents for 60 entries (Format is *Page: Date(with year-month-day). Title*)        
* [Page 1: 2018-01-24](#id-section1). Intro to Github, RMarkdown, and UNIX command-line
* [Page 2: 2018-01-29](#id-section2). Practice logging into UNIX and basic coding.
* [Page 3:2018-01-29](#id-section3).Working with RNA-Seq Data
* [Page 4:2018-01-31](#id-section4). Continuing to work with RNA-Seq Data
* [Page 5: 2018-02-05](#id-section5). Continuing to work with the RNA-Seq Data
* [Page 6:2018-02-07](#id-section6).Working with DESeq2
* [Page 7: 2018-02-12](#id-section7).Continuing to work with DESeq2. 
* [Page 8: 2018-02-14](#id-section8). FInishing work with DESeq2. 
* [Page 9: 2018-02021](#id-section9).DESeq2 one more time!
* [Page 10:](#id-section10).
* [Page 11:](#id-section11).
* [Page 12:](#id-section12).
* [Page 13:](#id-section13).
* [Page 14:](#id-section14).
* [Page 15:](#id-section15).
* [Page 16:](#id-section16).
* [Page 17:](#id-section17).
* [Page 18:](#id-section18).
* [Page 19:](#id-section19).
* [Page 20:](#id-section20).
* [Page 21:](#id-section21).
* [Page 22:](#id-section22).
* [Page 23:](#id-section23).
* [Page 24:](#id-section24).
* [Page 25:](#id-section25).
* [Page 26:](#id-section26).
* [Page 27:](#id-section27).
* [Page 28:](#id-section28).
* [Page 29:](#id-section29).
* [Page 30:](#id-section30).
* [Page 31:](#id-section31).
* [Page 32:](#id-section32).
* [Page 33:](#id-section33).
* [Page 34:](#id-section34).
* [Page 35:](#id-section35).
* [Page 36:](#id-section36).
* [Page 37:](#id-section37).
* [Page 38:](#id-section38).
* [Page 39:](#id-section39).
* [Page 40:](#id-section40).
* [Page 41:](#id-section41).
* [Page 42:](#id-section42).
* [Page 43:](#id-section43).
* [Page 44:](#id-section44).
* [Page 45:](#id-section45).
* [Page 46:](#id-section46).
* [Page 47:](#id-section47).
* [Page 48:](#id-section48).
* [Page 49:](#id-section49).
* [Page 50:](#id-section50).
* [Page 51:](#id-section51).
* [Page 52:](#id-section52).
* [Page 53:](#id-section53).
* [Page 54:](#id-section54).
* [Page 55:](#id-section55).
* [Page 56:](#id-section56).
* [Page 57:](#id-section57).
* [Page 58:](#id-section58).
* [Page 59:](#id-section59).
* [Page 60:](#id-section60).

------
<div id='id-section1'/>
### Page 1: 2018-01-24. Notes on using Github, Rmarkdown, and the UNIX command-line

Today, we created our github repos for the course, and began our notebooks. 

Other goals for today: 
* publish our notebooks
* log into the UNIX server


------
<div id='id-section2'/>
### Page 2: 2018-01-29. Practice logging into UNIX and basic coding.

Logged into UNIX server using Putty, accessed mydata, viewed cols_data.txt
------
<div id='id-section3'/>
### Page 3: 2018-01-29. Working with RNA-Seq Data. 

Goals: 
1. Learn about the bull headed beetle (*Orthophagus taurus*)
2. Understand pipeline for processing RNA-Seq data
3. Learn how to write bash scripts and write scripts to process files in batches
4. Visualize and interpret Illumina data quality

**Bull headed beetle
- Native to mediterranean, deliberately introduced to Australia for pest control
- Accidentally introduced to eastern US from unknown origin in 1970s
- Experimental design
*** Three populations reared in common garden experiment, from native range, Italy (IT), western australia (WA), and North Carolina (NC)
*** Four developmental stages, both sexes (three individuals per sex)
*** 3 pops x 4 developmental stages x 2 sexes x 3 individuals = 72 samples
*** Sequenced on about 7 lanes of Illumina HiSeq 2500

** The pipeline
1. Visualize, Clean, Visualize
2. Download reference transcriptome assembly
3. Map (aka align) cleaned reads from each sample to the reference assembly to generate sequence alignemnt files (Program: bwa, Input: fastq, Output: .sam)
4. Extract read count data from .sam files
5. Assemble a data matrix of counts for each gene for each sample

***Visualize,Clean,Visualize
- Access shared directory through pbio381 server (cd /data/project_data/beetles/rawdata/)
- Working with WA_PP1_M1* (zcat WA_PP1_M1_CCGTCC_L003_R1_001.fastq.gz | head -n4)
- Quality score: @@@DDDDDA?DCFEGGEHHHGGEGGII+C1?EHBHGG##0:BDDFGIIDHG1=CGIGGGEHB3?D>CCAA;?:>>?BA@CB@A>ACCCBBB#+8?BBC##
- Clean file using: fastqc WA_PP1_M1_CCGTCC_L003_R1_001.fastq.gz -o ~/mydata
- Move html file generated to computer using WINSCP
- Opened fastq file in html document

------
<div id='id-section4'/>
### Page 4: 2018-01-31. Continuing to work with RNA-Seq Data

1. Make directories "scripts" and "cleanreads"
2. Copy bash script over to ~/scripts directory 

```
cp /data/scripts/trim_example.sh ~/scripts/ # copies the script to your home scripts dir
vim trim_example.sh                       # open the script with vim to edit
```

```
#!/bin/bash
      java -classpath /data/popgen/Trimmomatic-0.33/trimmomatic-0.33.jar org.usadellab.trimmomatic.TrimmomaticPE \  
                -threads 1 \ 
                -phred33 \
                 /data/project_data/beetles/rawdata/WA_PP1_M1_CCGTCC_L003_R1_00_R1.fastq.gz \
                 /data/project_data/beetles/rawdata/WA_PP1_M1_CCGTCC_L003_R1_00_R2.fastq.gz \
                 ~/cleanreads/"WA_PP1_M1_CCGTCC_L003_R1_001_clean_paired.fa" \
                 ~/cleanreads/"WA_PP1_M1_CCGTCC_L003_R1_001_clean_unpaired.fa" \
                 ~/cleanreads/"WA_PP1_M1_CCGTCC_L003_R1_002_clean_paired.fa" \
                 ~/cleanreads/"WA_PP1_M1_CCGTCC_L003_R1_002_clean_unpaired.fa" \
                 ILLUMINACLIP:/data/popgen/Trimmomatic-0.33/adapters/TruSeq3-PE.fa:2:30:10 \
                 LEADING:28 \
                TRAILING:28 \
                SLIDINGWINDOW:6:28 \
                HEADCROP:12 \
                MINLEN:35 \
```
Now run the script
 ```
 chmod u+x trim_example.sh    # makes the script "executable" by the "user"
./trim_example.sh           # executes the script, or bash trim_example.sh
```
Output: 
```
TrimmomaticPE: Started with arguments: -threads 1 -phred33 /data/project_data/beetles/rawdata/WA_PP1_M1_CCGTCC_L003_R1_001.fastq.gz /data/project_data/beetles/rawdata/WA_PP1_M1_CCGTCC_L003_R2_001.fastq.gz /users/m/s/msouthga/cleanreads/WA_PP1_M1_CCGTCC_L003_R1_clean-paired.fa /users/m/s/msouthga/cleanreads/WA_PP1_M1_CCGTCC_L003_R1_clean_unpaired.fa /users/m/s/msouthga/cleanreads/WA_PP1_M1_CCGTCC_L003_R2_clean_paired.fa /users/m/s/msouthga/cleanreads/WA_PP1_M1_CCGTCC_L003_R2_clean_unpaired.fa ILLUMINACLIP:/data/popgen/Trimmomatic-0.33/adapters/TruSeq3-PE.fa:2:30:10 LEADING:28 TRAILING:28 SLIDINGWINDOW:6:28 HEADCROP:12 MINLEN:35
Using PrefixPair: 'TACACTCTTTCCCTACACGACGCTCTTCCGATCT' and 'GTGACTGGAGTTCAGACGTGTGCTCTTCCGATCT'
ILLUMINACLIP: Using 1 prefix pairs, 0 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences
```
Run FASTQC on our clean and paired reads, and store in fastqc directory
```
fastqc ~/cleanreads/WA_PP1_M1_CCGTCC_L003_R1_clean-paired.fa -o ~/fastqc/

```
Map your reads to the reference transcriptome
- First step is to index the reference transcriptome - we've done the step already, but here's the command: 

```
$ bwa index /data/project_data/beetles/reference/OTAU.fna
```
- make a new directory sam in mydata
- Our first step is to map the clean reads using the bwa mem command
```
bwa mem /data/project_data/beetles/reference/OTAU.fna ~/cleanreads/WA_PP1_M1_CCGTCC_L003_R1_clean-paired.fa ~/cleanreads/WA_PP1_M1_CCGTCC_L003_R2_clean_paired.fa > WA_PP1_M1_bwamem.sam

```




------
<div id='id-section5'/>
### Page 5: Continuing to work with RNA-Seq Data
navigate to mydata/sam
```
 tail -n 100 WA_PP1_M1_bwamem.sam > tail.sam   #take 100 last columns in file
 vim tail.sam   #view tail.sam file
 : set nowrap
```
A SAM file is a tab delimited text file that stores information about the alignment of reads in a FASTQ file to a reference gnome or transcriptome. For each read in a FASTQ file, there's a line in the SAM file that includes: - the read (aka query name)
- a FLAG (number with info about mapping success and orientation and whether the read is the left or right read)
 * Check FLAG scores using BROAD institue page (https://broadinstitute.github.io/picard/explain-flags.html)
- The reference sequence name to which the read mapped (if not mapped gets a *)
- The mapping quality (Phred-scaled)
- a CIGAR string that gives alignment information (how many bases Match, location of Insertion or Deletion)
-an '=', mate position, inferred insert size(columns 7,8,9)
- the query sequence and Phred-scaled quality from the FASTQ file (columsn 10 and 11)
- 

```
grep -c XT:A:U WA_PP1_M1_bwamem.sam  #use general expressions to search for specific term

samtools flagstat WA_PP1_M1_bwamem.sam

cp /data/scripts/countxpression_PE.py ~/myscripts  #move countxpression_PE.py file from shared directory to myscripts

vim countxpression_PE.py  # visualize file

```

------
<div id='id-section6'/>
### Page 6: 2018-02-07. Working with DESeq2. 
Working with R on personal machine using: 
- counts matrix
-compiled table of uniquely mapped reads to each gene from each sample

DESeq2 - can rerun with bwa align, get python script to work with mem, or choose different way. 
1) start with bwa align. Use WinSCP to transfer file into EcologicalGenomics/UNIXData/Feb7

work recorded in Feb7_DESeq2.R
------
<div id='id-section7'/>
### Page 7: Continuing to work with DESeq2. 
------
<div id='id-section8'/>
### Page 8: Finishing work with DESeq2. 

- Made a heatmap comparing expression levels
- Look at individual genes by devstage and sex using PCA

-GO enrichment using MWU test
------
<div id='id-section9'/>
### Page 9: One more day with DESeq2!
Goals: 
-1) Explore how we can use DESeq2 and other packages in R to test for differences in gene expression of different models
3) To identify co-expression modules using WCGNA

- IT population didn't provide all developmental stages - so missing from some of this analysis

- made heat map of significant expression differences between WA_PP1_F and NC_PP1_F (also with corresponding male samples)

- WGNCA = weighted gene network correlation analysis, performed using package WGNCA. 

_ brightest clusters and strongest correlations shown in developmental stage
------
<div id='id-section10'/>
### Page 10:
------
<div id='id-section11'/>
### Page 11:
------
<div id='id-section12'/>
### Page 12:
------
<div id='id-section13'/>
### Page 13:
------
<div id='id-section14'/>
### Page 14:
------
<div id='id-section15'/>
### Page 15:
------
<div id='id-section16'/>
### Page 16:
------
<div id='id-section17'/>
### Page 17:
------
<div id='id-section18'/>
### Page 18:
------
<div id='id-section19'/>
### Page 19:
------
<div id='id-section20'/>
### Page 20:
------
<div id='id-section21'/>
### Page 21:
------
<div id='id-section22'/>
### Page 22:
------
<div id='id-section23'/>
### Page 23:
------
<div id='id-section24'/>
### Page 24:
------
<div id='id-section25'/>
### Page 25:
------
<div id='id-section26'/>
### Page 26:
------
<div id='id-section27'/>
### Page 27:
------
<div id='id-section28'/>
### Page 28:
------
<div id='id-section29'/>
### Page 29:
------
<div id='id-section30'/>
### Page 30:
------
<div id='id-section31'/>
### Page 31:
------
<div id='id-section32'/>
### Page 32:
------
<div id='id-section33'/>
### Page 33:
------
<div id='id-section34'/>
### Page 34:
------
<div id='id-section35'/>
### Page 35:
------
<div id='id-section36'/>
### Page 36:
------
<div id='id-section37'/>
### Page 37:
------
<div id='id-section38'/>
### Page 38:
------
<div id='id-section39'/>
### Page 39:
------
<div id='id-section40'/>
### Page 40:
------
<div id='id-section41'/>
### Page 41:
------
<div id='id-section42'/>
### Page 42:
------
<div id='id-section43'/>
### Page 43:
------
<div id='id-section44'/>
### Page 44:
------
<div id='id-section45'/>
### Page 45:
------
<div id='id-section46'/>
### Page 46:
------
<div id='id-section47'/>
### Page 47:
------
<div id='id-section48'/>
### Page 48:
------
<div id='id-section49'/>
### Page 49:
------
<div id='id-section50'/>
### Page 50:
------
<div id='id-section51'/>
### Page 51:
------
<div id='id-section52'/>
### Page 52:
------
<div id='id-section53'/>
### Page 53:
------
<div id='id-section54'/>
### Page 54:
------
<div id='id-section55'/>
### Page 55:
------
<div id='id-section56'/>
### Page 56:
------
<div id='id-section57'/>
### Page 57:
------
<div id='id-section58'/>
### Page 58:
------
<div id='id-section59'/>
### Page 59:
------
<div id='id-section60'/>
### Page 60:

------